neo/rawio/blackrockrawio.py

```diff
@@ -853,354 +853,549 @@ def _extract_nev_file_spec(self):
 
-        filesize_bytes = self._get_file_size(filename)
+        # Create file memmap
+        file_memmap = np.memmap(filename, dtype='uint8', mode='r')
 
-        data_header = {}
-        if offset is None:
-            offset_to_first_data_block = int(self._nsx_basic_header[nsx_nb]["bytes_in_headers"])
-        else:
-            offset_to_first_data_block = int(offset)
+        # Get file parameters
+        filesize = self._get_file_size(filename)
+        channels = int(self._nsx_basic_header[nsx_nb]["channel_count"])
+        current_offset = int(self._nsx_basic_header[nsx_nb]["bytes_in_headers"])
 
-        channel_count = int(self._nsx_basic_header[nsx_nb]["channel_count"])
-        current_offset_bytes = offset_to_first_data_block
-        data_block_index = 0
-        while current_offset_bytes < filesize_bytes:
-            packet_header = self._read_nsx_dataheader(nsx_nb, current_offset_bytes)
-            header_flag = packet_header["header_flag"]
-            # NSX data blocks must have header_flag = 1, other values indicate file corruption
-            if header_flag != 1:
+        data_blocks = {}
+        block_idx = 0
+
+        # Loop through file, reading block headers
+        while current_offset < filesize:
+            # Read header at current position
+            header = self._read_nsx_dataheader(spec, nsx_nb, current_offset)
+
+            if header["header_flag"] != 1:
                 raise ValueError(
-                    f"Invalid NSX data block header at offset {current_offset_bytes:#x} in ns{nsx_nb} file. "
-                    f"Expected header_flag=1, got {header_flag}. "
-                    f"This may indicate file corruption or unsupported NSX format variant. "
-                    f"Block index: {data_block_index}, File size: {filesize_bytes} bytes"
+                    f"Invalid NSX data block header at offset {current_offset:#x} "
+                    f"in ns{nsx_nb} file. Expected header_flag=1, got {header['header_flag']}."
                 )
-            timestamp = packet_header["timestamp"]
-            num_data_points = int(packet_header["nb_data_points"])
-            offset_to_data_block_start = current_offset_bytes + packet_header.dtype.itemsize
-
-            data_header[data_block_index] = {
-                "header": header_flag,
-                "timestamp": timestamp,
-                "nb_data_points": num_data_points,
-                "offset_to_data_block": offset_to_data_block_start,
-            }
 
-            # Jump to the next data block, the data is encoded as int16
-            data_block_size_bytes = num_data_points * channel_count * np.dtype("int16").itemsize
-            current_offset_bytes = offset_to_data_block_start + data_block_size_bytes
+            num_samples = int(header["nb_data_points"])
+            data_offset = current_offset + header.dtype.itemsize
+            timestamp = header["timestamp"]
+
+            # Create data view into memmap for this block
```

Comment:
Here, we use views on a linear memmaped buffer to avoid creating a memmap per data block. This is good because OSes usually limit the ammount of memamps you can create.